### PR TITLE
Reproducing grudzien2020

### PR DESCRIPTION
**Pull request by cgrudz**
Re #70 I'm not sure that my new commits made it into my old pull request -- if this is redundant we can kill this request.  In any case, I have an updated example for you to take a look at.

### CODE REVIEW

File: examples/perfect_random_model_example.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+# #### Generate the same random numbers each time this script is run
+
+seed = dpr.set_seed(3000)
+
+# #### Simulate synthetic truth (xx) and noisy obs (yy)
+xx, yy = HMM_truth.simulate()
+
+# note - the truth twin takes twice the number of the integration steps
+# as the model twin in this configuration
+xx = xx[0::2, :]
```
Comment (patricknraanes) [inline examples/perfect_random_model_example.py:29]: This is the solution I would have suggested too :-)

One minor suggestion:

```suggestion
xx = xx[::2]
```

File: dapper/mods/integration.py [hunk 2]
```diff
@@ -12,8 +12,13 @@
...
 # fmt: off
-def rk4(f, x, t, dt, order=4):
-    """Runge-Kutta (explicit, non-adaptive) numerical ODE solvers.
+def rk4(f, x, t, dt, stages=4, s=0):
+    """Runge-Kutta (explicit, non-adaptive) numerical (S)ODE solvers.
+
+    The rule has strong / weak convergence order 1.0 for generic SDEs and order 4.0
+    convergence for ODEs when stages=4.  For stages=1, this becomes the Euler-Maruyama
+    schemefor SDEs (s > 0.0) with strong / weak convergence order 1.0 for SDEs with
+    additive noise as defined in the below. See `bib.grudzien2020numerical`.
```
Comment (patricknraanes) [inline dapper/mods/integration.py:21]: - Split deterministic and stochastic description into 2 paragraphs.
- Add reference for stochastic scheme.
- "Rule" --> "Scheme"

File: dapper/mods/integration.py [hunk 3]
```diff
@@ -29,38 +34,73 @@ def rk4(f, x, t, dt, order=4):
     dt : float
         Integration time step.
 
-    order : int, optional
-        The order of RK method. Default: 4
+    stages : int, optional
+        The number of stages of the RK method.
+        When stages=1, this becomes the Euler (-Maruyama) scheme.
```
Comment (patricknraanes) [inline dapper/mods/integration.py:39]:
```suggestion
        When `stages=1`, this becomes the Euler (-Maruyama) scheme.
```

**Review by patricknraanes — CHANGES_REQUESTED** [review]
I will just close the other PR anyway.

Hint: you can apply the suggested changes (if you're fine with them) in the GitHub interface. Probably you want to add them to a single batch before committing (still in the GitHub interface). The you can pull from origin to get those changes locally. Then you can add your own changes locally (for example to comments without suggested changes) and push, which should update this PR.

**Comment by patricknraanes** [comment]
I have made the document generation automatic. So please remove all html files listed under "Conflicting files" right below. Also, there is a conflict in `dapper/mods/integration.py` since I just made some changes there. In order to resolve, [pull from upstream](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/syncing-a-fork#syncing-a-fork-from-the-command-line), fix conflicts, and push to origin (which will automatically update this PR).

File: dapper/mods/L96s/Grudzien2020.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+####################
+# Main definitions #
+####################
+
+
+# -------------------------------#
+# define benchmark configuration #
+# -------------------------------#
+
```
[the remark this answers is not in the document]
Comment (cgrudz) [reply dapper/mods/L96s/Grudzien2020.py:32]: This is still a work in progress -- more sub-headings to come! hahahaha

File: dapper/mods/L96s/Grudzien2020.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
+###############
+# Description #
+###############
+
```
[the remark this answers is not in the document]
Comment (cgrudz) [reply dapper/mods/L96s/Grudzien2020.py:4]: Same as above, I'm just trying to commit some incremental changes for testing etc., more functionality and documentation to come.

File: examples/perfect_random_model_example.py (hunk 1 above)
Comment (cgrudz) [reply examples/perfect_random_model_example.py:29]: Cool, I don't remember the Python syntax that well, if you say that is the correct slicing let's do it.

File: dapper/mods/integration.py (hunk 2 above)
Comment (cgrudz) [reply dapper/mods/integration.py:21]: Sounds good -- I'll take a whack at that in a bit.

**Comment by cgrudz** [comment]
> I have made the document generation automatic. So please remove all html files listed under "Conflicting files" right below. Also, there is a conflict in `dapper/mods/integration.py` since I just made some changes there. In order to resolve, [pull from upstream](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/syncing-a-fork#syncing-a-fork-from-the-command-line), fix conflicts, and push to origin (which will automatically update this PR).

Nice, will do.  I have plans to finish off some of this merger today, along with *hopefully* having time after meetings to build some exercises and further examples.  Take a look at the example if you got a chance -- the simulation is looking good, confirming the results of the GMD work.

**Comment by patricknraanes** [comment]
Noice!

dpr

**Comment by cgrudz** [comment]
Yeah, I think it should be fun to play with some of these time series, especially tuning the amount of stochasticity.  It begins to look more and more like a Brownian motion with enough noise, or more like the standard Lorenz 96 in the small noise, there should be some nice exercises to look at in an Jupyter notebook.

**Comment by patricknraanes** [comment]
I think the model should be named `Lorenz96s` which lines up logically with the other models
Screenshot from 2021-07-16 19-41-13

Also, please de-capitalize Grudzien2020. Nobody else gets capitalization :laughing:

**Comment by cgrudz** [comment]
Hahaha, yeah, I have to wait until I have my own toy model until I get capitalization....

**Comment by cgrudz** [comment]
Hey, I thought I merged the changes you wanted from integrate, but now I get this commit error
dapper/mods/integration.py
    94:  18 F821 undefined name 'functools' [

**Comment by patricknraanes** [comment]
> Hey, I thought I merged the changes you wanted from integrate, but now I get this commit error
> dapper/mods/integration.py
> 94: 18 F821 undefined name 'functools' [

I's `functools` somewhere in the file? Line 94?

**Comment by cgrudz** [comment]
Sorry my bad, incorrectly was merging changes.  Let's see how the new pushes look.

**Comment by patricknraanes** [comment]
The tests are crashing coz it still says `L96s` in `grudzien2020`

**Comment by patricknraanes** [comment]
Sorry, I am on holiday and am not active. How's this looking? AFAICT there's still some suggested changes and comments have not been addressed.

**Comment by cgrudz** [comment]
No worries, getting back to finishing some of this off tomorrow, just 
juggling this with finishing off some other loose ends.

File: dapper/mods/L96s/Grudzien2020.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+# -------------------------------#
+# define benchmark configuration #
+# -------------------------------#
+
+# Grudzien 2020 uses the below chronology with KObs=25000, BurnIn=5000
+ttruth = modelling.Chronology(dt=0.005, dtObs=0.1, T=30, Tplot=Tplot, BurnIn=10)
+tmodel = modelling.Chronology(dt=0.01, dtObs=0.1, T=30, Tplot=Tplot, BurnIn=10)
+
+# set the system diffusion coefficient
+diff = 0.10
```
[the remark this answers is not in the document]
Comment (cgrudz) [reply dapper/mods/L96s/Grudzien2020.py]: Sounds good, for sure, I'll try to propagate this throughout any instances.

File: dapper/mods/integration.py (hunk 2 above)
Comment (cgrudz) [reply dapper/mods/integration.py]: You'll need to see if you like the table for the orders of convergence, etc... see most recent commits.

File: dapper/mods/L96s/__init__.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    # draw standard normal sample to define the
+    # recursive Stratonovich integral coefficients
+    rndm = np.random.standard_normal([5, sys_dim])
+    xi = rndm[0, :]
+
+    mu = rndm[1, :]
+    phi = rndm[2, :]
+
+    zeta = rndm[3, :]
+    eta = rndm[4, :]
```
[the remark this answers is not in the document]
Comment (cgrudz) [reply dapper/mods/L96s/__init__.py]: Tested and merged.

File: dapper/mods/L96s/Grudzien2020.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+# set the model state vector dimension
+Nx = 10
+
+# define the initial condition
+x0 = x0(Nx)
+
+
+# define different numerical step models for the ensemble, to vary the precision
+def em_ensemble_step(x0, t, dt):
+    # Euler-Maruyama (order 1.0 Weak / Strong)
```
[the remark this answers is not in the document]
Comment (cgrudz) [reply dapper/mods/L96s/Grudzien2020.py]: Do you have some time to chat about these steppers this week?  I think I might be running into a structural issue with DAPPER that I want to discuss, as there may be an easy implementation that I don't see.

File: dapper/mods/integration.py [hunk 9]
```diff
@@ -10,8 +10,29 @@
...
+    |      3 |            3.0  |                  N/A |                    N/A |
+    ----------------------------------------------------------------------------
+    |      4 |            4.0  |                  1.0 |                    1.0 |
+    ----------------------------------------------------------------------------
+
+    For stages=1, this becomes the Euler(-Maruyama) schemefor SDEs (s > 0.0)
+    with strong / weak convergence order 1.0 for SDEs with additive noise, see
+    `bib.grudzien2020numerical` and references therein for a general discussion
+    of numerical integegration schemes and discretization error analysis in the
+    DA cycle.
```
Comment (patricknraanes) [inline dapper/mods/integration.py:35]: I really do like tables, but I think in this case it's not needed. Also the reduction to Euler's scheme is mentioned under the `stages` keyword (below).
```suggestion
    """Runge-Kutta (explicit, non-adaptive) numerical (S)ODE solvers.

    For ODEs, the order of convergence equals the number of `stages`.
    For SDEs with additive noise (`s>0`), the order of convergence
    (both weak and strong) is 1 for any number of `stages`.
    See `bib.grudzien2020numerical` for a DA-specific discussion on
    integration schemes and their discretization errors.
```

**Review by patricknraanes — CHANGES_REQUESTED** [review]

**Comment by patricknraanes** [comment]
Apart from the suggested changes in `rk4` everything looks great now.

**Comment by patricknraanes** [comment]
I just upgraded the documentation for [`setup`](https://nansencenter.github.io/DAPPER/dapper/xp_launch.html#dapper.xp_launch.run_experiment) quite a bit. Please have a look.

Also added to `bib.py`, unfortunately causing conflict. Please pull, resolve, push.

I also think you should mainly build on `basic_2`, not `basic_3`. Because although `basic_3` illustrates the use of `setup` it also contains other complications. So start with `basic_2`. Use a simple for-loop to generate several `xp`s, not the methods shown in `basic_3`. Sorry I don't have time right now to look in more detail.

**Comment by cgrudz** [comment]
Don't worry, I'm probably not going to touch this again this week -- 
most of the template for the replication based on basic_3 is already 
done though, I just need to make some plotting scripts.

**Comment by patricknraanes** [comment]
Yeah, it just seems to me like there might be some confusion still, and it is much easier to consult on this when the script follows the template of `basic_2`, without the complications of `basic_3`. Maybe just stash the plotting stuff for later?

**Comment by cgrudz** [comment]
No worries, I had meetings cancelled and had extra time today, got this 
debugged and fully reproducing results from the manuscript now.  The 
README is updated accordingly, along with some additional references 
that I think are useful... including my own Julia code base hehehe.  
I'll work on plotting probably next week.

File: dapper/mods/Lorenz96s/__init__.py [hunk 10]
```diff
@@ -131,17 +131,17 @@ def Psi(l1, l2):
 def em_ensemble_step(x0, t, dt):
     """Euler-Maruyama (order 1.0 Weak / Strong) model step wrapper"""
 
-    return rk4(lambda t, x: dxdt(x), x0, np.nan, dt, s=diffusion, stages=1)
+    return rk4(lambda t, x: dxdt(x), x0, np.nan, dt, s=Diffusion, stages=1)
```
Comment (patricknraanes) [inline dapper/mods/Lorenz96s/__init__.py:134]: Capital `D` in variable name is rather unconventional for Python

**Comment by cgrudz** [comment]
Will probably spend some time on plotting scripts later this week, minor changes have addressed items above.  If you have any extra docs or comments to push to the plotting example based on basic 3, let me know, otherwise I'll just try tinkering with this more.

**Comment by cgrudz** [comment]
In fact, I'm getting bogged down a bit on the plotting code, I'm thinking to remove the draft that I have currently, push the change and go for a merger of branches for now.  Does that sound good?  I'll get back to this again shortly but I'll need a bit of lead time as I'm prepping for the semester.

**Pull request merged**